HPCC-13663 Implement a stable merge sort

ghalliday opened the pull request:
Initial pull request for review.
The commits are separate logical stages, and may be worth reviewing separately.

This was tested sorting 284M rows in hthor, with memory and cpu bound to the same node.  The median times of 5 runs were as follows:

Original: 6m22s
Commit1: 6m14s  Returns rows sorted in place rather than using an index. Slightly surprised this is faster
Commit2: 4m03    First version of merge sort
Commit3: 4m01.2  Optimized version

Commit4, Commit 5:  4m01.1
I was slightly surprised, but the last two commits slightly improve the performance, but possibly not by enough to be worth including.  Especially the 3way compare which adds complexity.

The merge sort example is now spending 50% of its time in the compare function, 20% allocating and freeing rows.  I suspect it is close to optimum, but any bright ideas welcome.

ghalliday committed:
HPCC-13663 Refactor stable quicksort to return sorted rows in place

Previously the function returned an sorted array of pointers to rows
which is an unusual interface and is harder to optimize with another
function.
Also remove the sort with two compare functions which was unused.

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
HPCC-13663 Implementation of a vector merge sort

Allow merge and parallel quick sort to be configured in hthor to
allow comparison testing.

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday commented:
@richardkchapman @jakesmith @mckellyln please review.

A few Q's:
* Should we default to always using this for stable sorts (currently only hthor is calling it)
* Should I remove the 3-way optimization, since it is is quite a lot of code for little benefit.

I am starting to work on a parallel implementation - I think I might need to implement a task manager as part of it, so it may take a little while.

richardkchapman commented:
How does it do on pathogenic cases (already sorted, already reverse sorted, all rows the same, etc) ?

richardkchapman commented:
To answer your questions:

1. We should add it as a new selectable sort algorithm. If we a confident it is always better than QuickSort for stable, we should use it as the default when no algorithm selected. Can it slot into the spilling sort algorithm too?
2. If I'm reading your timings correctly, the last two commits together made a difference of 0.1 seconds out of 240 seconds - that seems too small to be significant to me, and they certainly render the code harder to follow/review for correctness.

jakesmith commented on system/jlib/jsort.cpp:
```diff
@@ -678,6 +608,178 @@ void parqsortvecstable(void ** const rows, size32_t n, const ICompare & compare,
+{
+    if (n <= 1)
+        return;
+    mergeSort(rows, n, compare, temp, 0);
```
Ignore my previous comment that's in one of the commits referring to this

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

ghalliday commented:
@richardkchapman I've cleaned up the code, removed the cacheline and 3 element compare, and added options to hthor to allow the sort algorithm to be selected.

I would plan to allow the algorithm to be selected in roxie in a later pull request.  I also wonder if we should support ,PARALLEL as a flag on SORT (and anything else that allows the algorithm to be supplied).

"How does it do on pathogenic cases (already sorted, already reverse sorted, all rows the same, etc) ?"

There aren't really any pathogenic cases  - each of those would execute slightly faster than the average case (since in each case above all rows would be copied from one partition first, therefore it would almost halve the number of compares).

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK
